### PR TITLE
Basic parsing of dependencies and phases

### PR DESCRIPTION
**Pull request by richfitz**
~Merge after #6, contains those commits~

This PR does the next tranche of multi-statement validation and preparation moving through the system and working out if we have a clear path through the dependency tree, and which equations will be used in which parts of the system (handling parameters, computing the rhs or update function, doing a comparison to data, etc).

There's another little bit that will follow that I might roll into the next stage (beginning of generation) which is working out where we physically store the data.

### CODE REVIEW

File: R/dependencies.R [hunk 1]
```diff
@@ -23,3 +23,30 @@ find_dependencies <- function(expr) {
        variables = unique(variables$get()))
 
 }
+
+
```
Comment (weshinsley) [inline R/dependencies.R:27]: So here, deps is a list of named vectors;
each vector has the name of a var, and the value is a vector of vars that it depends on.
deps is produced by parse_system_depends

File: R/dependencies.R [hunk 2]
```diff
@@ -23,3 +23,30 @@ find_dependencies <- function(expr) {
...
+topological_order <- function(deps) {
+  if (all(lengths(deps) == 0)) {
+    return(list(success = TRUE, value = seq_along(deps)))
+  }
+
```
Comment (weshinsley) [inline R/dependencies.R:32]: convert deps into a matrix of ndeps * ndeps - where m[x,y] - being TRUE implies the variable at index x is depended on, by the variable at index y. So in the test example:

a = c+1, b = a*2, c = b/2:

deps contanins $a = "c", $b = "a", $c = "b"

m[3,1] is true, (c is depended on by a), as are m[1,2] and m[2,3]

File: R/dependencies.R [hunk 3]
```diff
@@ -23,3 +23,30 @@ find_dependencies <- function(expr) {
...
+  m <- matrix(FALSE, length(deps), length(deps))
+  for (i in seq_along(deps)) {
+    m[, i] <- unname(names(deps) %in% deps[[i]])
+  }
+
+  pending <- rep(TRUE, length(deps))
+  ret <- integer(0)
+  while (any(pending)) {
+    i <- which(pending)[colSums(m[, pending, drop = FALSE]) == 0]
```
Comment (weshinsley) [inline R/dependencies.R:41]: Here we find entries where m[1,1] = m[2,1] = m[3,1] = FALSE - meaning a does not depend on anything else. 
If we find such entries, then 'a' is known, so set m[1,1], m[1,2], m[1,3] to FALSE and don't worry about 'a' any more.

Then try again and see if we can resolve anymore - until hopefully we have no unresolved vars. 

But if we try, and can't do this to any of the vars (so `i` comes back empty) then we are cyclic and can't resolve.

File: R/dependencies.R [hunk 4]
```diff
@@ -23,3 +23,30 @@ find_dependencies <- function(expr) {
...
+    if (length(i) > 0L) {
+      ret <- c(ret, i)
+      pending[i] <- FALSE
+      m[i, ] <- FALSE
+    } else {
+      return(list(success = FALSE, error = which(pending)))
```
Comment (weshinsley) [inline R/dependencies.R:47]: If we fail, return the indexes into deps of the vars we couldn't resolve.

**Review by weshinsley — APPROVED** [review]

**Review by weshinsley — APPROVED** [review]
Looks good - I wonder whether a bit more internal documentation/commenting might be good - I left my train of thought through the toplogy function which I had to stare at a little before it properly made sense.

**Comment by richfitz** [comment]
yeah, feel free to document the algorithm in a comment if you'd like.  I've shuttled a version of that algorithm around since first writing it in about 2014!

**Pull request merged**